File: kt/aoc2023-lib/src/commonMain/kotlin/com/github/ephemient/aoc2023/Day1.kt
package com.github.ephemient.aoc2023

class Day1(input: String) {
    private val lines = input.lines()

    fun part1() = solve(digits)

    fun part2() = solve(extendedDigits)

    private fun solve(values: Collection<IndexedValue<String>>): Int = lines.sumOf { line ->
        val subset = values.filter { it.value in line }
        if (subset.isNotEmpty()) {
            10 * subset.minBy { line.indexOf(it.value) }.index + subset.maxBy { line.lastIndexOf(it.value) }.index
        } else {
            0
        }
    }

    companion object {
        private val digits = List(10) { IndexedValue(it, it.toString()) }
        private val extendedDigits = digits + listOf(
            IndexedValue(1, "one"),
            IndexedValue(2, "two"),
            IndexedValue(3, "three"),
            IndexedValue(4, "four"),
            IndexedValue(5, "five"),
            IndexedValue(6, "six"),
            IndexedValue(7, "seven"),
            IndexedValue(8, "eight"),
            IndexedValue(9, "nine"),
        )
    }
}

File: py/aoc2023/day1.py
"""
Day 1: Trebuchet?!
"""

SAMPLE_INPUT_1 = """
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""
SAMPLE_INPUT_2 = """
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"""


def _solve(data, mapping):
    acc = 0
    for line in data.splitlines():
        subset = {key: value for key, value in mapping.items() if key in line}
        if not subset:
            continue
        # pylint: disable=cell-var-from-loop
        x = min(subset.items(), key=lambda item: line.index(item[0]))[1]
        y = max(subset.items(), key=lambda item: line.rindex(item[0]))[1]
        acc += 10 * x + y
    return acc


DIGITS = {str(d): d for d in range(10)}
EXTENDED_DIGITS = DIGITS | {
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
}


def part1(data):
    """
    >>> part1(SAMPLE_INPUT_1)
    142
    """
    return _solve(data, DIGITS)


def part2(data):
    """
    >>> part2(SAMPLE_INPUT_2)
    281
    """
    return _solve(data, EXTENDED_DIGITS)


parts = (part1, part2)
